File: Ruby/Algo2.rb
require 'benchmark'

def get_mode(mod)
    frequency = {}

    bgst_freq = 1

    mode_vals = []

    for i in 0...mod.length 
        num = mod[i]

        if frequency[num]
            frequency[num] += 1
        else 
            frequency[num] = 1
        end

        if frequency[num] > bgst_freq
            bgst_freq = frequency[num]
        end
    end

    frequency.each do |num, freq|
        if freq == bgst_freq
            mode_vals.push(num)
        end
    end

    return mode_vals
end

result = Benchmark.measure do
    numsArray = []

    File.open("./data/50_nums.txt", "r") do |file|
      content = file.read
      numsArray = content
                    .gsub("[", "") # Remove the opening barcket
                    .gsub("]", "")
                    .split(",")
                    .map { |num| num.strip.to_f } # Removes spaces and converts to float
    end

    puts    " \n Vals mode: #{get_mode(numsArray)}"
end

puts " \n Execution time: #{result.real} milliseconds"

File: Python/Algo2.py
import time

# Measuring total time including the output
start_time = time.time()

# Read the file and process the numbers
with open("./data/1000_nums.txt", "r") as file:
    content = file.read()
    nums_array = (
        content.replace("[", "")  # Remove the opening bracket
               .replace("]", "")  # Remove the closing bracket
               .split(",")        # Split the numbers by comma
    )
    nums_array = [float(num.strip()) for num in nums_array]  # Remove spaces and convert to float

def get_mode(mod):
    frequency = {}
    bgst_freq = 1
    mode_vals = []

    for num in mod:
        if num in frequency:
            frequency[num] += 1
        else:
            frequency[num] = 1

        if frequency[num] > bgst_freq:
            bgst_freq = frequency[num]

    for num in frequency:
        if frequency[num] == bgst_freq:
            mode_vals.append(num)

    return mode_vals

mode_result = get_mode(nums_array)
end_time = time.time()

total_time = (end_time - start_time) * 1000  # Convert to milliseconds

print('Mode:', mode_result, f"\n Executed in: {total_time:.4f} ms")
